File: backend/services/furniture_service.py
import uuid
from datetime import datetime
from database import db
from models.furniture import Furniture, FurnitureLog
from services.file_service import upload_file

def create_furniture(property_id, name, status="Good", purchase_price=0.0, note=None, image=None):
    """
    Create a furniture item and optionally upload an image.
    Folder: properties/{prop_id}/furnitures/{fur_id}/img
    """
    try:
        # 1. Create DB Record first to get the ID
        new_item = Furniture.create(
            property_id=property_id,
            name=name,
            status=status,
            purchase_price=purchase_price,
            note=note
        )

        # 2. Handle Image Upload if provided
        if image:
            folder_name = f"properties/{property_id}/furnitures/{new_item.id}/img"
            ext = image.filename.rsplit('.', 1)[-1].lower()
            filename = f"{uuid.uuid4()}.{ext}"

            image_url = upload_file(image, folder_name, filename)
            
            # Update record with URL
            new_item.image_url = image_url
            db.session.commit()

        return True, "Furniture created successfully", new_item

    except Exception as e:
        db.session.rollback()
        return False, str(e), None

def update_furniture(furniture_id, data, image=None):
    """
    Update furniture details.
    """
    item = Furniture.find_by_id(furniture_id)
    if not item:
        return False, "Furniture not found"

    # Update fields if they exist in data
    if "name" in data: item.name = data["name"]
    if "status" in data: item.status = data["status"]
    if "purchase_price" in data: item.purchase_price = float(data["purchase_price"])
    if "note" in data: item.note = data["note"]

    # Handle Image Update
    if image:
        folder_name = f"properties/{item.property_id}/furnitures/{item.id}/img"
        ext = image.filename.rsplit('.', 1)[-1].lower()
        filename = f"{uuid.uuid4()}.{ext}"

        image_url = upload_file(image, folder_name, filename)
        item.image_url = image_url

    db.session.commit()
    return True, "Furniture updated successfully"

def delete_furniture(furniture_id):
    """
    Delete furniture and all associated logs (Cascade delete handled by DB model usually, 
    but we explicitly commit here).
    """
    item = Furniture.find_by_id(furniture_id)
    if not item:
        return False, "Furniture not found"

    db.session.delete(item)
    db.session.commit()
    return True, "Furniture deleted"

def get_furniture_by_property(property_id):
    items = Furniture.find_by_property_id(property_id)

    result = []

    for i in items:
        result.append({
            "id": i.id,
            "property_id": property_id,
            "name": i.name,
            "status": i.status,
            "purchase_price": i.purchase_price,
            "image_url": i.image_url,
            "note": i.note,
            "added_date": i.added_date.isoformat() if i.added_date else None
        })


    return result



def add_log(furniture_id, log_type, description, date=None, image=None):
    """
    Add a log and perform Smart Actions on furniture status.
    Folder: properties/{prop_id}/furnitures/{fur_id}/logs/{log_id}/img
    """
    item = Furniture.find_by_id(furniture_id)
    if not item:
        return False, "Furniture not found", None

    if item.status == "Disposed":
        return False, "Cannot add logs to a disposed item", None

    try:
        if log_type == "Repair":
            item.status = "Good"
        elif log_type == "Damage":
            item.status = "Damaged"
        elif log_type == "Dispose":
            item.status = "Disposed"

        log_date = datetime.strptime(date, '%Y-%m-%d') if date else datetime.now(datetime.timezone.utc)
        
        new_log = FurnitureLog.create(
            furniture_id=furniture_id,
            log_type=log_type,
            description=description,
            date=log_date
        )

        if image:
            folder_name = f"properties/{item.property_id}/furnitures/{item.id}/logs/{new_log.id}/img"
            ext = image.filename.rsplit('.', 1)[-1].lower()
            filename = f"{uuid.uuid4()}.{ext}"

            image_url = upload_file(image, folder_name, filename)
            new_log.image_url = image_url
            
        db.session.commit()
        return True, "Log added successfully", new_log

    except Exception as e:
        db.session.rollback()
        return False, str(e), None

def delete_log(log_id):
    log = FurnitureLog.query.get(log_id)
    if not log:
        return False, "Log not found"

    db.session.delete(log)
    db.session.commit()
    return True, "Log deleted"

def get_furniture_details(furniture_id):
    """
    Get a specific furniture item AND its logs.
    """

    item = Furniture.find_by_id(furniture_id)
    if not item:
        return None

    logs = FurnitureLog.find_by_furniture_id(furniture_id)

    log_list = []
    for l in logs:
        log_list.append({
            "id": l.id,
            "log_type": l.log_type,
            "description": l.description,
            "date": l.date.isoformat(),
            "image_url": l.image_url
        })

    data = {
        "id": item.id,
        "property_id": item.property_id,
        "name": item.name,
        "status": item.status,
        "purchase_price": item.purchase_price,
        "image_url": item.image_url,
        "note": item.note,
        "added_date": item.added_date.isoformat() if item.added_date else None,
        "logs": log_list  
    }

    return data

File: flutter_application/lib/services/furniture_service.dart
import 'dart:convert';
import 'package:flutter/cupertino.dart';
import 'package:http/http.dart' as http;
import 'package:image_picker/image_picker.dart';
import 'package:flutter_application/services/api_service.dart';
import 'package:flutter_application/models/furniture.dart';

class FurnitureService {


  static Future<bool> createFurniture({
    required int propertyId,
    required String name,
    String status = "Good",
    double purchasePrice = 0.0,
    String? note,
    XFile? image,
  }) async {
    final uri = ApiService.buildUri("/furniture/create");

    try {
      var request = http.MultipartRequest('POST', uri);

      request.fields['property_id'] = propertyId.toString();
      request.fields['name'] = name;
      request.fields['status'] = status;
      request.fields['purchase_price'] = purchasePrice.toString();
      if (note != null) request.fields['note'] = note;

      if (image != null) {
        final bytes = await image.readAsBytes();
        request.files.add(
          http.MultipartFile.fromBytes(
            'image',
            bytes,
            filename: image.name,
          ),
        );
      }

      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);
      final jsonData = jsonDecode(response.body);

      if (response.statusCode == 201 && jsonData['success'] == true) {
        return true;
      } else {
        debugPrint("Create furniture failed: ${jsonData['message']}");
        return false;
      }
    } catch (e) {
      debugPrint("Exception during create furniture: $e");
      return false;
    }
  }

  static Future<bool> updateFurniture({
    required int furnitureId,
    required Map<String, dynamic> fields, 
    XFile? image,
  }) async {
    final uri = ApiService.buildUri("/furniture/update");

    try {
      var request = http.MultipartRequest('POST', uri);

      request.fields['furniture_id'] = furnitureId.toString();

      fields.forEach((key, value) {
        if (value != null) {
          request.fields[key] = value.toString();
        }
      });

      if (image != null) {
        final bytes = await image.readAsBytes();
        request.files.add(
          http.MultipartFile.fromBytes(
            'image',
            bytes,
            filename: image.name,
          ),
        );
      }

      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);
      final jsonData = jsonDecode(response.body);

      if (response.statusCode == 200 && jsonData['success'] == true) {
        return true;
      } else {
        debugPrint("Update furniture failed: ${jsonData['message']}");
        return false;
      }
    } catch (e) {
      debugPrint("Exception during update furniture: $e");
      return false;
    }
  }

  static Future<bool> deleteFurniture(int furnitureId) async {
    final uri = ApiService.buildUri("/furniture/delete");

    try {
      final response = await http.post(
        uri,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'furniture_id': furnitureId}),
      );

      final jsonData = jsonDecode(response.body);

      if (response.statusCode == 200 && jsonData['success'] == true) {
        return true;
      } else {
        debugPrint("Delete furniture failed: ${jsonData['message']}");
        return false;
      }
    } catch (e) {
      debugPrint("Exception during delete furniture: $e");
      return false;
    }
  }

  static Future<List<Furniture>> getFurnitureByProperty(int propertyId) async {
    final uri = ApiService.buildUri("/furniture/list/$propertyId");

    try {
      final response = await http.get(uri);
      final jsonData = jsonDecode(response.body);
  

      if (response.statusCode == 200 && jsonData['success'] == true) {
          
        List<dynamic> list = jsonData['data'];
        return list.map((item) => Furniture.fromJson(item)).toList();
      } else {
        debugPrint("Get furniture list failed: ${jsonData['message']}");
        return [];
      }
    } catch (e) {
      debugPrint("Exception fetching furniture list: $e");
      return [];
    }
  }

  static Future<Furniture?> getFurnitureDetails(int furnitureId) async {
    final uri = ApiService.buildUri("/furniture/$furnitureId");

    try {
      final response = await http.get(uri);
      final jsonData = jsonDecode(response.body);

      if (response.statusCode == 200 && jsonData['success'] == true) {
        return Furniture.fromJson(jsonData['data']);
      } else {
        debugPrint("Get furniture details failed: ${jsonData['message']}");
        return null;
      }
    } catch (e) {
      debugPrint("Exception fetching furniture details: $e");
      return null;
    }
  }

  static Future<bool> addLog({
    required int furnitureId,
    required String logType, 
    required String description,
    required DateTime date,
    XFile? image,
  }) async {
    final uri = ApiService.buildUri("/furniture/log/add");

    try {
      var request = http.MultipartRequest('POST', uri);

      request.fields['furniture_id'] = furnitureId.toString();
      request.fields['log_type'] = logType;
      request.fields['description'] = description;
      request.fields['date'] = date.toIso8601String().split('T')[0]; 

      if (image != null) {
        final bytes = await image.readAsBytes();
        request.files.add(
          http.MultipartFile.fromBytes(
            'image',
            bytes,
            filename: image.name,
          ),
        );
      }

      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);
      final jsonData = jsonDecode(response.body);

      if (response.statusCode == 201 && jsonData['success'] == true) {
        return true;
      } else {
        debugPrint("Add log failed: ${jsonData['message']}");
        return false;
      }
    } catch (e) {
      debugPrint("Exception during add log: $e");
      return false;
    }
  }

  static Future<bool> deleteLog(int logId) async {
    final uri = ApiService.buildUri("/furniture/log/delete");

    try {
      final response = await http.post(
        uri,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'log_id': logId}),
      );

      final jsonData = jsonDecode(response.body);

      if (response.statusCode == 200 && jsonData['success'] == true) {
        return true;
      } else {
        debugPrint("Delete log failed: ${jsonData['message']}");
        return false;
      }
    } catch (e) {
      debugPrint("Exception during delete log: $e");
      return false;
    }
  }
}